helper_test.go

```diff
@@ -1924,27 +1925,60 @@ func randomSemver(t *testing.T) string {
 	return fmt.Sprintf("%d.%d.%d", rand.Intn(99)+3, rand.Intn(99)+1, rand.Intn(99)+1)
 }
 
+var testNames map[string]int = nil
+
 func checkTestNodeEnv(t *testing.T) {
-	if nodeIndex, ok := os.LookupEnv("CI_NODE_INDEX"); ok {
-		testChar := strings.ToLower(t.Name())[4]
-		testIndex := getTestIndex(testChar)
-		if testIndex != nodeIndex {
-			t.Skip()
+	t.Helper()
+
+	if nodeIndex, ok := envLookupInteger(t, "CI_NODE_INDEX"); ok {
+		if nodeTotal, ok := envLookupInteger(t, "CI_NODE_TOTAL"); ok {
+			if testNames == nil {
+				testNames = listTestsCI(t)
+			}
+
+			testIndex, ok := testNames[t.Name()]
+			if !ok {
+				t.Fatalf("%s was not found in the list of tests.", t.Name())
+			}
+
+			if testIndex%nodeTotal != nodeIndex {
+				t.Skip()
+			}
 		}
 	}
 }
 
-func getTestIndex(x byte) string {
-	if x >= 'a' && x <= 'h' {
-		return "0"
+func envLookupInteger(t *testing.T, name string) (int, bool) {
+	raw, ok := os.LookupEnv(name)
+	if !ok {
+		return 0, false
 	}
-	if x >= 'i' && x <= 'p' {
-		return "1"
+
+	result, err := strconv.Atoi(raw)
+	if err != nil {
+		t.Fatalf("Expected %s to be an integer, but got %s", name, raw)
 	}
-	if x >= 'q' && x <= 'u' {
-		return "2"
+	return result, true
+}
+
+func listTestsCI(t *testing.T) map[string]int {
+	t.Helper()
+
+	cmd := exec.Command("go", "test", "./...", "-list=.", "-tags=integration")
+	output, err := cmd.Output()
+	if err != nil {
+		t.Fatalf("Failed to list all test files. Are you using go1.19+?: %v", err)
+	}
+
+	result := make(map[string]int)
+	index := 0
+	for _, line := range strings.Split(string(output), "\n") {
+		if strings.HasPrefix(line, "Test") {
+			result[line] = index
+			index += 1
+		}
 	}
-	return "3"
+	return result
```

Comment:
I just ran `go test -v ./... -list=. -tags=integration` and it is a tad bit slow. Alternatively, we can add the assignment for `testNames` to `init()` so it runs before any tests are run: 
```go
func init() {
  testNames = listTestsCI()
}
```
and remove `t.Helper()`. What do you think?

Reply:
To be clear I should only be running it once per node because it's being assigned to a package variable

Reply:
I think init is a better approach. I'll modify it.